src/lib/libparse/print.5
.TH print_control 5 "4 March, 1991" "Psi Release 2.0" "\*(]D"
.SH NAME
prcntr, prup, prdup \- control printing in a Fortran 77 program

.SH SYNOPSIS
.nf
integer function prcntr(request)
character*(*) request
#include <error.h>

integer function prup(keyword)
character*(*) keyword
#include <error.h>

integer function prdup()
#include <error.h>
.fi

.SH DESCRIPTION
.LP
These Fortran
routines provide a method of specifying print options in the input.
Print options are associated with symbolic names and can be turned on or
off by the user or by the program.  A program using
these print routines can easily determine whether
a given print option is on or off.  To use these routines the program
must be linked with the Psi input parsing library (libparse.a).  The
include file, error.h, can be found in the Psi include directory.

.LP
The \fBprcntr\fP function is used to turn print options on, off, and to
determine whether or not a print option has been turned on.
Calling \fBprcntr\fP with a argument of the
form '\fBON\fP \fIoption1\fP ... \fIoptionN\fP' turns on the listed
options.  Again, a return code of EOK (defined as 0 in error.h)
indicates that the call successfully completed.
Calling \fBprcntr\fP with a argument of the
form '\fBOFF\fP \fIoption1\fP ... \fIoptionN\fP' turns off the listed
options.  In this case, a return code of EOK
indicates that the call successfully completed.  If one of the print
options was off to begin with, then that option is ignored.
Calling \fBprcntr\fP with a argument of the
form '\fBIS_ON\fP \fIoption\fP' queries the given
option.  It will return 1 if the option is on and 0 if the option is off or
unknown.

.LP
The \fBprup\fP function is used to read in user defined print options from the
input.  It reads from the \fIkeyword\fP parameter in the input.
The form of the input is a vector of print lists.
Each print list is itself a vector of the form \fB(\fP\fIcontrol
option1 ... optionN\fP\fB)\fP, where \fIcontrol\fP is either \fBON\fP
or \fBOFF\fP and each \fIoption\fP is the print option affect.  The
vectors are read from the left to the right, so the
input 
.B "((ON OPTIONX OPTIONY) (OFF OPTIONX))"
has the effect of turning \fBOPTIONX\fP off and \fBOPTIONY\fP on.

.LP
The \fBprdup\fP sets up the default user print options.  It first
calls \fBprup(':DEFAULT:PRINT')\fP and then
calls \fBprup(':\fP\fIprogram_identifier\fP\fB:PRINT')\fP.

.LP
Psi programmers should follow the convention that if the print
option \fBBRIEF\fP is turned on, then only essential information
is included in the output.

.LP
Psi programmers should not use the print option \fBKEYWORD\fP.
This is reserved for use by the input parsing library.



src/lib/libparse/print.F
c Print option control routines.
c The names of print options are truncated to 8 characters.
c-----------------------------------------------------------------------------
c  Printing utility control routine
c This routine processes all print option functions
c  req is:
c    IS_ON followed by a print option names
c    ON followed one or more print option names
c    OFF followed by one or more print option names
c Error recovery: returns a value from error.h
c if the IS_ON request is given, a 0 or 1 is returned for true or false
      integer function prcntr(req)
#ifdef sun
      implicit undefined (a-z)
#else
      implicit integer (a-z)
#endif
#include "error.h"
#include "params.h"
      character*(*) req

      integer i
      integer maxprt
      parameter (maxprt = 100)
      character*80 next
      character*8 word1, opt
      save names
      save nnames
      integer nnames,ndelet
      character*8 names(maxprt)
      integer errcod
      character*80 emess
      common /lparer/ errcod, emess

      integer chrlen

      data nnames/0/

c Default return action is EOK
      errcod = EOK
      prcntr = errcod
      emess = ' '

      word1 = req(1:chrlen(req))
      next  = req(chrlen(req)+1:len(req))
      call leftju(next)

      if ((len(req).gt.80).and.(word1.ne.'IS_ON')) then
        errcod = Elimit
        prcntr = errcod
        emess = 'Request string is too long (prcntr)'
        return
        endif

      if (word1.eq.'IS_ON') then
        do 101 i = 1, nnames
          if (names(i).eq.next) then
            prcntr = 1
            return
            endif
  101     continue
        prcntr = 0
        return
      else if ((word1.eq.'ON').or.(word1.eq.'OFF')) then

c       Loop thru all of the names in the next list
 801    continue
          opt = next(1:chrlen(next))
          next  = next(chrlen(next)+1:len(next))
          call leftju(next)
          if (opt.eq.' ') return
          if (word1.eq.'ON') then
c           If an option is given twice, ignore it.
            do 102 i = 1, nnames
              if (names(i).eq.opt) return
 102          continue
c           Bounds checking
            if (nnames.ge.maxprt) then
              errcod = Elimit
              prcntr = errcod
              emess = 'Too many print option names (prcntr)'
              return
              endif
            nnames = nnames + 1
            names(nnames) = opt
          else
c           This will delete all occurrances of the opt
            ndelet = 0
            do 103 i = 1, nnames
              names(i-ndelet) = names(i)
              if (names(i).eq.opt) ndelet = ndelet + 1
  103         continue
            nnames = nnames - ndelet
            endif
          goto 801
      else
        errcod = Earg
        prcntr = errcod
        emess = 'Bad request, '''
        call scat(emess,word1)
        call scat(emess,''' (prcntr)')
        return
        endif
c Execution cannot reach here
      end
c---------------------------------------------------------------------------
c  The Printing utility user print routine
c This finds the array of character specified by the keywrd argument.
c The form of the vector is as follows:
c  ((ON NAME_1 NAME_2 ...)
c   (OFF NAME_1 NAME_3 ... ))
c Printing is turned off for names following OFF.
c Printing is turned on for names following ON.
      integer function prup(keywrd)
#ifdef sun
      implicit undefined (a-z)
#else
      implicit integer (a-z)
#endif
#include "params.h"
#include "error.h"
      character*(*) keywrd

      character*(lentok) opt,val,req
      integer errcod
      character*80 emess
      common /lparer/ errcod, emess

      integer frdpos, frdlpn, frdchr, prcntr

      prup = frdpos(keywrd)
      if (prup.ne.EOK) return

      prup = frdlpn()
      if (prup.ne.EOK) return

 801  continue
        prup = frdchr(val)
        if (prup.eq.Ebvect) then
          prup = frdchr(opt)
          if (prup.eq.Eevect) then
            goto 801
          else if (prup.ne.EOK) then
            return
            endif
 802      continue
            prup = frdchr(val)
            if (prup.ne.Eevect) then
              if (prup.ne.EOK) return
              req = opt
              call scatb(req,val)
              prup = prcntr(req)
              if (prup.ne.EOK) return
              goto 802
              endif
          goto 801
        else if (prup.eq.Eevect) then
          errcod = EOK
          prup = errcod
          emess = ' '
        else if (prup.eq.EOK) then
          errcod = Esyntx
          prup = errcod
          emess = 'Looking for a '')'' or ''('' but found a'''
          call scat(emess, val)
          call scat(emess, ''' (prup)')
          endif
      return
      end
c---------------------------------------------------------------------------
c  The Printing utility default user print routine
c This lets an application easily allow users to specify print options.
c This searches the current inputs for the PRINT_ON and PRINT_OFF keywords.
c   First the PRINT keyword from DEFAULT is examined.
c   Third the PRINT keyword from PROGID is examined.
      integer function prdup()
#ifdef sun
      implicit undefined (a-z)
#else
      implicit integer (a-z)
#endif
#include "params.h"
#include "error.h"

      character*(lentok) progid,keywrd
      integer errcod
      character*80 emess
      common /lparer/ errcod, emess

      integer prup

c Get the input from the defaults section
      keywrd = ':DEFAULT'
      call scat(keywrd,':PRINT')
      prdup = prup(keywrd)
      if (prdup.eq.EEOD) then
        errcod = EOK
        prdup = errcod
        emess = ' '
        endif
      if (prdup.ne.EOK) return

c Now get input for the programid
      call gprgid(progid)
      keywrd = ':'
      call scat(keywrd,progid)
      call scat(keywrd,':PRINT')
      prdup = prup(keywrd)
      if (prdup.eq.EEOD) then
        errcod = EOK
        prdup = errcod
        emess = ' '
        endif
      if (prdup.ne.EOK) return
      end

